examples/parser/Parser/Core.roc

```diff
@@ -0,0 +1,292 @@
+## `run` or `Parser.Str.runStr` etc. are more useful in daily usage.
+parsePartial : Parser input a, input -> ParseResult input a
+parsePartial = \@Parser parser, input ->
+    parser input
+
+## Runs a parser on the given input, expecting it to fully consume the input
+##
+## The `input -> Bool` parameter is used to check whether parsing has 'completed',
+## (in other words: Whether all of the input has been consumed.)
+##
+## For most (but not all!) input types, a parsing run that leaves some unparsed input behind
+## should be considered an error.
+parse : Parser input a, input, (input -> Bool) -> Result a [ParsingFailure Str, ParsingIncomplete input]
+parse = \parser, input, isParsingCompleted ->
+    when parsePartial parser input is
+        Ok { val: val, input: leftover } ->
+            if isParsingCompleted leftover then
+                Ok val
+            else
+                Err (ParsingIncomplete leftover)
+
+        Err (ParsingFailure msg) ->
+            Err (ParsingFailure msg)
+
+## Parser that can never succeed, regardless of the given input.
+## It will always fail with the given error message.
+##
+## This is mostly useful as 'base case' if all other parsers
+## in a `oneOf` or `alt` have failed, to provide some more descriptive error message.
+fail : Str -> Parser * *
+fail = \msg ->
+    buildPrimitiveParser \_input -> Err (ParsingFailure msg)
+
+## Parser that will always produce the given `val`, without looking at the actual input.
+## This is useful as basic building block, especially in combination with
+## `map` and `apply`.
+const : a -> Parser * a
+const = \val ->
+    buildPrimitiveParser \input ->
+        Ok { val: val, input: input }
+
+## Try the `first` parser and (only) if it fails, try the `second` parser as fallback.
+alt : Parser input a, Parser input a -> Parser input a
+alt = \first, second ->
+    buildPrimitiveParser \input ->
+        when parsePartial first input is
+            Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+            Err (ParsingFailure firstErr) ->
+                when parsePartial second input is
+                    Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+                    Err (ParsingFailure secondErr) ->
+                        Err (ParsingFailure ("\(firstErr) or \(secondErr)"))
+
+## Runs a parser building a function, then a parser building a value,
+## and finally returns the result of calling the function with the value.
+##
+## This is useful if you are building up a structure that requires more parameters
+## than there are variants of `map`, `map2`, `map3` etc. for.
+##
+## For instance, the following two are the same:
+##
+## >>> const (\x, y, z -> Triple x y z)
+## >>> |> map3 Parser.Str.nat Parser.Str.nat Parser.Str.nat
+##
+## >>> const (\x -> \y -> \z -> Triple x y z)
+## >>> |> apply Parser.Str.nat
+## >>> |> apply Parser.Str.nat
+## >>> |> apply Parser.Str.nat
```

Comment:
Is it worth mentioning why someone might choose `apply` over `mapN`, besides `apply` providing arbitrary-arity? IN general should one always prefer to use a `mapN` variant? If not why do `mapN` exist?